src/Bicep.LangServer/Registry/ModuleRestoreScheduler.cs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Registry;
+using Bicep.Core.Syntax;
+using Bicep.LanguageServer.CompilationManager;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bicep.LanguageServer.Registry
+{
+    public sealed class ModuleRestoreScheduler : IModuleRestoreScheduler, IDisposable
+    {
+        private record QueueItem(ICompilationManager CompilationManager, DocumentUri Uri, ImmutableArray<ModuleDeclarationSyntax> References);
+
+        private record CompletionNotification(ICompilationManager CompilationManager, DocumentUri Uri);
+
+        private readonly IModuleRegistryDispatcher dispatcher;
+
+        private readonly Queue<QueueItem> queue = new();
+
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+
+        // block on initial wait until signaled
+        private readonly ManualResetEventSlim manualResetEvent = new(false);
+
+        private bool disposed = false;
+        private Task? consumerTask;
+
+        public ModuleRestoreScheduler(IModuleRegistryDispatcher dispatcher)
+        {
+            this.dispatcher = dispatcher;
+        }
+
+        /// <summary>
+        /// Requests that the specified modules be restored to the local file system.
+        /// Does not wait for the operation to complete and returns immediately.
+        /// </summary>
+        /// <param name="references">The module references</param>
+        public void RequestModuleRestore(ICompilationManager compilationManager, DocumentUri documentUri, IEnumerable<ModuleDeclarationSyntax> references)
+        {
+            this.CheckDisposed();
+            var item = new QueueItem(compilationManager, documentUri, references.ToImmutableArray());
+            lock (this.queue)
+            {
+                this.queue.Enqueue(item);
+
+                // notify consumer about new items
+                this.manualResetEvent.Set();
+            }
+        }
+
+        public void Start()
+        {
+            this.CheckDisposed();
+            this.consumerTask = Task.Factory.StartNew(this.ProcessQueueItems, this.cancellationTokenSource.Token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+        }
+
+        public void Dispose()
+        {
+            // this is a sealed class - no need for full IDisposable implementation
+            if(!this.disposed)
+            {
+                this.disposed = true;
+                if (this.consumerTask is not null)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    this.consumerTask = null;
```

Comment:
Should you wait for the consumer task to exit here? It may be unexpected that a disposed class is still doing stuff outside of a `using` block.

Reply:
It actually causes a deadlock if we do `.Wait()` (at least in tests). Yeah, it's definitely unexpected, but the language server process shuts down immediately afterwards, so I'm not too worried about it being imperfect.

Reply:
> It actually causes a deadlock if we do `.Wait()` (at least in tests).

That feels like something we should try to understand IMO. Either that, or avoid declaring support for `IDisposable`. If we end up in a situation where we leave the background task running without termination, will that prevent app shutdown?

Reply:
No, left-over running tasks don't prevent app shutdown, but I will switch over to `DisposeAsync()` to fix the deadlock. The DI container supports `IAsyncDisposable`, so it should still be ok. (It's the DI container that will call us in the language server.)